internal/directives/json_updater.go

```diff
@@ -0,0 +1,166 @@
+package directives
+
+import (
+	"context"
+	"fmt"
+	"os"
+	"slices"
+	"strings"
+
+	securejoin "github.com/cyphar/filepath-securejoin"
+	"github.com/tidwall/sjson"
+	"github.com/xeipuuv/gojsonschema"
+
+	kargoapi "github.com/akuity/kargo/api/v1alpha1"
+)
+
+func init() {
+	builtins.RegisterPromotionStepRunner(newJSONUpdater(), nil)
+}
+
+// jsonUpdater is an implementation of the PromotionStepRunner interface that
+// updates the values of specified keys in a JSON file.
+type jsonUpdater struct {
+	schemaLoader gojsonschema.JSONLoader
+}
+
+// newJSONUpdater returns an implementation of the PromotionStepRunner interface
+// that updates the values of specified keys in a JSON file.
+func newJSONUpdater() PromotionStepRunner {
+	r := &jsonUpdater{}
+	r.schemaLoader = getConfigSchemaLoader(r.Name())
+	return r
+}
+
+// Name implements the PromotionStepRunner interface.
+func (j *jsonUpdater) Name() string {
+	return "json-update"
+}
+
+// RunPromotionStep implements the PromotionStepRunner interface.
+func (j *jsonUpdater) RunPromotionStep(
+	ctx context.Context,
+	stepCtx *PromotionStepContext,
+) (PromotionStepResult, error) {
+	failure := PromotionStepResult{Status: kargoapi.PromotionPhaseErrored}
+
+	if err := j.validate(stepCtx.Config); err != nil {
+		return failure, err
+	}
+
+	// Convert the configuration into a typed struct
+	cfg, err := ConfigToStruct[JSONUpdateConfig](stepCtx.Config)
+	if err != nil {
+		return failure, fmt.Errorf("could not convert config into %s config: %w", j.Name(), err)
+	}
+
+	return j.runPromotionStep(ctx, stepCtx, cfg)
+}
+
+// validate validates jsonUpdater configuration against a JSON schema.
+func (j *jsonUpdater) validate(cfg Config) error {
+	return validate(j.schemaLoader, gojsonschema.NewGoLoader(cfg), j.Name())
+}
+
+func (j *jsonUpdater) runPromotionStep(
+	_ context.Context,
+	stepCtx *PromotionStepContext,
+	cfg JSONUpdateConfig,
+) (PromotionStepResult, error) {
+	updates := make(map[string]any, len(cfg.Updates))
+	for _, update := range cfg.Updates {
+		var value any
+		if update.Value != nil {
+			switch {
+			case update.Value.String != "":
+				value = update.Value.String
+			case len(update.Value.AnythingArray) > 0:
+				value = update.Value.AnythingArray
+			case len(update.Value.AnythingMap) > 0:
+				value = update.Value.AnythingMap
+			case update.Value.Double != 0:
+				value = update.Value.Double
```

Comment:
This means you can never set the value of a number field to 0.

This is a consequence of the quirky code generation, which we will need to work around somehow. It may require some experimentation.

[This comment](https://github.com/akuity/kargo/pull/3151/files#r1889056508) is related.

Reply:
See [this comment](https://github.com/akuity/kargo/pull/3151/files#r1889083798) also.

That should help.